internal/buffer/search.go

```diff
@@ -67,50 +67,27 @@ func (b *Buffer) findDown(r *regexp.Regexp, start, end Loc) ([2]Loc, bool) {
 
 	rPadded := padRegexp(r)
 
-	for i := start.Y; i <= end.Y; i++ {
-		l, charpos, padMode := findLineParams(b, start, end, i)
-
-		match := rPadded[padMode].FindIndex(l)
-
-		if match != nil {
-			start := Loc{charpos + util.RunePos(l, match[0]), i}
-			if padMode&padStart != 0 {
-				start = start.Move(1, b)
-			}
-			end := Loc{charpos + util.RunePos(l, match[1]), i}
-			if padMode&padEnd != 0 {
-				end = end.Move(-1, b)
-			}
-			return [2]Loc{start, end}, true
-		}
-	}
-	return [2]Loc{}, false
-}
-
-func (b *Buffer) findUp(r *regexp.Regexp, start, end Loc) ([2]Loc, bool) {
-	lastcn := util.CharacterCount(b.LineBytes(b.LinesNum() - 1))
-	if start.Y > b.LinesNum()-1 {
-		start.X = lastcn - 1
-	}
-	if end.Y > b.LinesNum()-1 {
-		end.X = lastcn
-	}
-	start.Y = util.Clamp(start.Y, 0, b.LinesNum()-1)
-	end.Y = util.Clamp(end.Y, 0, b.LinesNum()-1)
-
-	if start.GreaterThan(end) {
-		start, end = end, start
+	var i0, i1, is int
+	if down {
+		i0, i1, is = start.Y, end.Y, 1
+	} else {
+		i0, i1, is = end.Y, start.Y, -1
```

Comment:
Argh, this looks ugly... I missed the fact that the direction of the loop also differs, so deduplication is not quite as trivial as I thought.

I'd suggest to drop this for now, leaving this refactoring effort to another PR (if at all).

Reply:
OK. I've dropped the last commit and squashed the other ones to a single one. I think we're done now. Also thanks for your explanations about `start` and `end`.